18284: digital business card

amanji opened the pull request:
*Issue #:* bcgov/entity#18284

Target release: [Entities Sprint 19.3b Release - Digital Credentials](https://app.zenhub.com/workspaces/entities-team-space-6143567664fb320019b81f39/reports/release?release=Z2lkOi8vcmFwdG9yL1JlbGVhc2UvOTYzOTI)

*Description of changes:*


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the business-filings-ui license (Apache 2.0).

severinbeauvais commented on package.json:
Please update the app version (line 3) and run `npm i` again.

This seems like a significant change. Do you agree the major release number should be updated?

amanji replied:
At least a minor change. I don't think there are any breaking changes that require a major update. But I'm happy to do that.

severinbeauvais replied:
You're right, but I still think it's significant enough to earn a major release increase :)

severinbeauvais replied:
Please make it "7.0.0" :)

amanji replied:
Whoops. Thanks for catching that!

severinbeauvais replied:
Looks good.

severinbeauvais commented on src/components/dialogs/index.ts:
Do you think we should keep this feature more self-contained? I'm thinking you should move the dialogs to src/components/DigitalCredentials/. Thoughts?

severinbeauvais replied:
Changed. Thanks. Resolving this conversation.

severinbeauvais commented on src/enums/routes.ts:
```diff
@@ -13,7 +13,5 @@ export enum Routes {
-  DOWNLOAD_WALLET = 'download-wallet',
-  REGISTER_WALLET = 'register-wallet',
-  ISSUE_CREDENTIALS = 'issue-credentials'
+  ISSUE = 'issue'
```
Is this still a pilot? If not then please update line 14.

Could you come up with a less generic name for this route please? :)

severinbeauvais replied:
Thanks for changing the key name, but please also update the value (route name).

amanji replied:
The path is currently `/digitalCredentials/issue`. Would it be more descriptive to change it to `digitalCredentials/issue-credential`? Just seems like a long path name.

severinbeauvais replied:
Oh, I missed the fact that it's a child route... because it's declared like all other routes in the enum.

It makes sense to have 2 child routes ("" and "issue") but I'd like to prevent this confusion again. Ideas?

severinbeauvais replied:
Sounds good. Go for it.

severinbeauvais replied:
Looks good. Resolving.

severinbeauvais commented on src/interfaces/steps-interface.ts:
```diff
@@ -4,5 +4,6 @@ export interface StepsIF {
   icon: string
   text: string
   to: string,
+  disabled: boolean,
```
Is this property always used? If not then change to `disabled?: boolean`.

amanji replied:
I recall getting a TS error for this being optional but I'll take a look again.

severinbeauvais replied:
Looks good.

severinbeauvais commented on src/services/legal-services.ts:
```diff
@@ -89,9 +94,12 @@ export default class LegalServices {
+        .get(url)
+        // workaround because data is at "response.data.data"
+        .then((response) => response?.data)
+    )
```
I'm curious why you change the code style one way here and a different way above. Why not leave it as it was?

amanji replied:
ES lint auto formats the code. I can align the style with what's above.

severinbeauvais replied:
Oh, we've seen this before, but not recently.

Is this what your linting step looks like?

![image](https://github.com/bcgov/business-filings-ui/assets/10002889/97e94f05-f4a9-4da9-8162-6fed4dd6d7c3)

Do you perhaps have a different linter installed globally? Here's what I have:

![image](https://github.com/bcgov/business-filings-ui/assets/10002889/ba9248e9-1e06-4dbf-96ad-5d1b2fdbbf45)

severinbeauvais replied:
^^ When I run lint, it doesn't reformat anything.

amanji replied:
It might be my IDE settings, auto formatting. I have the Prettier extension installed which seems to be doing its thing on save. I will revert the un-related changes.

severinbeauvais replied:
Thanks for reverting. Looks good.

severinbeauvais commented:
Akiff, I can't remember what the plan was -- will we merge your code (UI and API) into main branch for now, then port it to Legal Name branch? Or should we merge this PR to Legal Name branch right away? It depends on when you need to be in Dev or Test environment (although this PR could have a preview site in GCP instead of its own environment).

severinbeauvais commented:
Akiff, can you look at the SonarCloud code smells, please?

amanji commented:
> Akiff, I can't remember what the plan was -- will we merge your code (UI and API) into main branch for now, then port it to Legal Name branch? Or should we merge this PR to Legal Name branch right away? It depends on when you need to be in Dev or Test environment (although this PR could have a preview site in GCP instead of its own environment).

I think for now we've decided to merge this into `main`. I've been advised that there could be some time before the Legal Name branch is ready. Perhaps for now we can just create the preview site in GCP if that will be faster. It's really just to allow internal folks to test out the application.

severinbeauvais commented:
> I think for now we've decided to merge this into `main`. I've been advised that there could be some time before the Legal Name branch is ready. Perhaps for now we can just create the preview site in GCP if that will be faster. It's really just to allow internal folks to test out the application.

This is fine for the UI. Do you have clarity on where the API changes will be merged? (You may need special config in the UI if it needs to talk to the Legal Name branch API.)

amanji commented:
> > I think for now we've decided to merge this into `main`. I've been advised that there could be some time before the Legal Name branch is ready. Perhaps for now we can just create the preview site in GCP if that will be faster. It's really just to allow internal folks to test out the application.
> 
> 
> 
> This is fine for the UI. Do you have clarity on where the API changes will be merged? (You may need special config in the UI if it needs to talk to the Legal Name branch API.)

I'll have to convene with Argus on that. I believe for now that would also be in `main`.

amanji commented:
@severinbeauvais I've addressed the code smells. The remaining failures are due to unrelated code changes.

severinbeauvais commented:
> @severinbeauvais I've addressed the code smells. The remaining failures are due to unrelated code changes.

Thanks. Don't worry about the code duplication warning.

severinbeauvais commented:
Akiff, the following command will build this PR to a preview site (and another message will let us know its URL when it's up). This command builds the CURRENT code, so you'll need to run it again if you change the code.

I'm not 100% that your code will run correctly right now because I don't know if you have any UI env key or FF changes, and the Legal API changes you need aren't deployed yet either. But it's a start :)

/gcbrun

bcregistry-sre commented:
Temporary Url for review: https://business-filings-dev--pr-552-zpnetclq.web.app

bcregistry-sre commented:
Temporary Url for review: https://business-filings-dev--pr-552-zpnetclq.web.app

severinbeauvais commented:
What should I see if I load the Sole Prop entity dashboard (Filings UI) at the preview site?

bcregistry-sre commented:
Temporary Url for review: https://business-filings-dev--pr-552-zpnetclq.web.app

severinbeauvais commented on src/components/DigitalCredentials/CredentialsDashboard.vue:
```diff
@@ -0,0 +1,153 @@
+    ConfirmRevokeCredentialDialog,
+    ConfirmReplaceCredentialDialog,
+    CredentialRevokedDialog,
+    ReplaceCredentialErrorDialog,
```
This dialog does not appear to be used. Future, or delete the references?

amanji replied:
Added it to the UI

severinbeauvais replied:
Looks good.

severinbeauvais commented on src/components/DigitalCredentials/CredentialsDashboard.vue:
```diff
@@ -0,0 +1,153 @@
+  replaceCredentialErrorDialog = false;
+  revokeCredentialErrorDialog = false;
+  issuedCredential: DigitalCredentialIF = null;
+  issuedCredentials: Array<DigitalCredentialIF> = [];
```
No need for trailing semi-colons. Please remove.

severinbeauvais replied:
Looks good.

severinbeauvais commented on src/components/DigitalCredentials/CredentialsDashboard.vue:
```diff
@@ -0,0 +1,153 @@
+
+  async getCredentials (): Promise<void> {
+    const { data } = await LegalServices.fetchCredentials(this.getIdentifier)
+    this.issuedCredentials = data?.issuedCredentials
```
Conditional chaining may result in an undefined value, which isn't reactive in Vue (and you use this variable as a prop) so I recommend assigning a fallback value (`|| []`).

severinbeauvais replied:
Looks good.

severinbeauvais commented on src/components/DigitalCredentials/CredentialsInfo.vue:
```diff
@@ -0,0 +1,70 @@
+      <v-card-text>
+        <p class="mb-2">
+          In the near future you will be able to use your business card credeptial to obtain a business licence,
+          business bank account, business permits. You can also get access to services like the forest client
```
"and business permits" ?

severinbeauvais replied:
Looks good.

severinbeauvais commented on src/components/DigitalCredentials/CredentialsInfo.vue:
```diff
@@ -0,0 +1,70 @@
+          class="text-decoration-none"
+          href=""
+          target="_blank"
+        >Signup to be part of the business licence credential pilot </a>
```
"Sign up" ?

Should any of these apps be proper nouns and capitalized?

Is this still called a pilot?

severinbeauvais replied:
Looks good.

severinbeauvais commented on src/components/DigitalCredentials/CredentialsInfo.vue:
```diff
@@ -0,0 +1,70 @@
+<template>
+  <div>
```
Would it make sense to use `aside` tag here?

severinbeauvais replied:
Not changing this. Resolving.

severinbeauvais commented on src/components/DigitalCredentials/CredentialsLanding.vue:
```diff
@@ -1,70 +1,54 @@
+              <!-- FIXME: Path aliases don't seem to be working -->
+              <!-- <v-img
+                        src="@/assets/images/digital_credentials.svg"
+                      /> -->
```
Might need an alias in Vite config?

severinbeauvais replied:
If images are now working (at preview site), please resolve this comment.

severinbeauvais commented on src/components/DigitalCredentials/CredentialsLanding.vue:
```diff
@@ -1,70 +1,54 @@
+                Business Card Credentials
+              </h1>
+              <p>
+                A digital credential is the digital equivalent of paper-based credential like a drivers license.
```
"of **a** paper-based"

"driver's license"

severinbeauvais replied:
Looks good.

severinbeauvais commented on src/components/DigitalCredentials/CredentialsDashboard.vue:
```diff
@@ -0,0 +1,153 @@
+<template>
+  <div>
```
I like to see ids on components to faciliate debugging and testing. Our standard is to name it after the component itself, eg, `id="credential-dashboard"`.

Also for the other components, and any major sections/blocks/v-cards/etc.

(Use classnames if the component can be displayed multiple times, to avoid duplicate ids.)

severinbeauvais replied:
Looks good in all components I checked. Thanks!

severinbeauvais commented:
@leodube-aot @JazzarKarim Have a look at this code change. I'm more looking for you to be aware of this than to deeply review it, please and thanks!

severinbeauvais commented on src/components/DigitalCredentials/CredentialsStepper.vue:
```diff
@@ -0,0 +1,237 @@
+                Steps to get your credentials:
+              </h1>
+              <v-card class="mb-2">
+                <v-card-text class="d-flex">
```
A v-card in a v-card is odd. They are usually used to display a white "card" on an otherwise grey page.

I also noticed that your v-cards have a shadow, which is different from our usual style. Please confirm this what your UI/UX team wants.

amanji replied:
This is what's indicated in the wireframes. Can we maintain it for now?

severinbeauvais replied:
Yes. Resolving.

severinbeauvais commented on src/components/DigitalCredentials/CredentialsStepper.vue:
```diff
@@ -0,0 +1,237 @@
+
+  credentialTypes = DigitalCredentialTypes;
+  credentialConnection: WalletConnectionIF = null;
+  issuedCredential: DigitalCredentialIF = null;
```
Our code style is to not have trailing semi-colons. Please remove. Thanks!

severinbeauvais replied:
Looks good.

severinbeauvais commented on src/components/DigitalCredentials/CredentialsStepper.vue:
```diff
@@ -0,0 +1,237 @@
+
+  async getCredentialsConnection (): Promise<void> {
+    const { data } = await LegalServices.fetchCredentialConnections(this.getIdentifier)
+    this.credentialConnection = data?.connections?.[0]
```
Conditional chaining can return undefined, which might make this variable un-responsive. I recommend a fallback value (eg, null).

severinbeauvais replied:
Looks good.

severinbeauvais commented on src/components/DigitalCredentials/CredentialsStepper.vue:
```diff
@@ -0,0 +1,237 @@
+<style lang="scss" scoped>
+.credentials-stepper-number {
+  color: rgb(255, 255, 255);
+  background-color: rgb(22, 105, 187);
```
For white, you can use "white".

For the other one, can you reuse one of the app colours defined in theme.scss?

severinbeauvais replied:
Looks good.

severinbeauvais commented on src/components/DigitalCredentials/CredentialsStepper.vue:
```diff
@@ -0,0 +1,237 @@
+.word-wrap {
+  word-break: normal;
+}
+</style>
```
Can you combine these style blocks?

For this particular class,  you may want to add "word-break-normal" in base.scss instead.

severinbeauvais replied:
Looks good.

severinbeauvais commented on src/components/DigitalCredentials/CredentialsTable.vue:
```diff
@@ -1,8 +1,9 @@
-    id="credentials-table"
-    class="mt-8"
-  >
+  <div>
```
Again, please set an ID for all components on the top-level element.

severinbeauvais replied:
Looks good.

severinbeauvais commented on src/components/DigitalCredentials/CredentialsTable.vue:
```diff
@@ -1,8 +1,9 @@
-    class="mt-8"
-  >
+  <div>
+    <h1 class="mb-4">
```
The coding style in this UI is mostly to declare top whitespace (instead of bottom whitespace or worse, both!). I'd prefer to see mt-4 on line 7 instead.

severinbeauvais replied:
Thanks for changing this. Looks good.

severinbeauvais commented on src/components/DigitalCredentials/CredentialsTable.vue:
```diff
@@ -21,25 +22,37 @@
+          <CredentialsMenu
+            :issuedCredential="item"
+            @onConfirmRevokeCredential="promptConfirmRevokeCredential"
+            @onConfirmReplaceCredential="promptConfirmReplaceCredential"
```
I prefer to see event handlers look like functions, eg
```
@onConfirmRevokeCredential="promptConfirmRevokeCredential($event)"
```

severinbeauvais replied:
Looks good.

severinbeauvais commented on src/components/DigitalCredentials/CredentialsTable.vue:
```diff
@@ -21,25 +22,37 @@
+  @Prop({ default: () => [] }) readonly issuedCredentials!: Array<DigitalCredentialIF>;
 
   get credentialsTableHeaders (): Array<TableHeaderIF> {
     // Do not display headers if there is table data
```
"if there is NO table data"

severinbeauvais replied:
Looks good.

severinbeauvais commented on src/components/DigitalCredentials/dialogs/ConfirmReplaceCredentialDialog.vue:
```diff
@@ -0,0 +1,127 @@
+}
+
+.warning-title {
+  font-size: 1.5rem;
```
For fonts, use the CSS variables in theme.scss. Here, it would be `$px-24`.

severinbeauvais replied:
For this and my other comments, if they apply to other files then please fix there also (so I don't have to repeat these comments). Thanks!

amanji replied:
Updated the theme file with additional sizes

severinbeauvais replied:
Looks good.

As mentioned elsewhere, I'm OK with using these "font" variables for line heights since they're related.

severinbeauvais reviewed (commented):
I think I've scanned/reviewed all files up to now. I'll review any new commits as they come in. Let me close my conversations as I use them as "bookmarks" for code that I want to re-review. Thanks.

amanji commented:
> I think I've scanned/reviewed all files up to now. I'll review any new commits as they come in. Let me close my conversations as I use them as "bookmarks" for code that I want to re-review. Thanks.

Thanks @severinbeauvais this is really helpful. I'll get through these.

amanji commented:
> @amanji , I change this PR's base to feature-digital-credentials branch (which was created from v6.9.2 code, which is what is currently in Prod, and which this PR was already based on). This is so we can release this branch independently from main branch (which contains other changes).

Sounds good. Working on addressing the code review comments.

severinbeauvais commented:
/gcbrun

severinbeauvais commented on src/assets/styles/base.scss:
Looks like lint had a heyday with the code again.

How much pain for you to revert this? Otherwise it's pain for the reviewers. lmk

severinbeauvais replied:
Actually, never mind -- there aren't that many code changes to review here.

amanji replied:
I will revert this. Sorry, also will disable auto formatting in my IDE

bcregistry-sre commented:
Temporary Url for review: https://business-filings-dev--pr-552-zpnetclq.web.app

severinbeauvais commented on src/components/DigitalCredentials/dialogs/CredentialRevokedDialog.vue:
```diff
@@ -63,26 +64,26 @@ export default class CredentialRevokedDialog extends Vue {
   color: $gray9;
   text-align: center;
-  line-height: 2.125rem;
+  line-height: $px-34;
```
Since line height is tied to font size, it makes sense to use these CSS variables. 👍

severinbeauvais commented:
Akiff, will you be updating EntityMenu.vue in this ticket to move the BDC menu into the dropdown?

The menu item is already protected by a FF so that's good, but the route itself (Credentials Dashboard) should also be protected so users can't bypass the menu and just enter the URL.

severinbeauvais commented:
> The menu item is already protected by a FF so that's good, but the route itself (Credentials Dashboard) should also be protected so users can't bypass the menu and just enter the URL.

To protect the Credentials Dashboard, you could do what the other filing views here do: programmatically pass an id on the route, and redirect to the app dashboard if it's missing.

Eg,
- pass an id: https://github.com/bcgov/business-filings-ui/blob/563a701703e0aaed6d6c0e55f975e684cb3329d2/src/components/EntityInfo/EntityMenu.vue#L314
- check for id: https://github.com/bcgov/business-filings-ui/blob/563a701703e0aaed6d6c0e55f975e684cb3329d2/src/views/AgmLocationChg.vue#L464

severinbeauvais commented:
/gcbrun

severinbeauvais commented:
Akiff, please fix the broken unit tests. It may be due to FFs... see src/utils/feature-flags.ts and tests/unit/EntityMenu.spec.ts in this PR for some fixes: https://github.com/bcgov/business-filings-ui/pull/558

bcregistry-sre commented:
Temporary Url for review: https://business-filings-dev--pr-552-zpnetclq.web.app

severinbeauvais commented on src/components/DigitalCredentials/CredentialsMenu.vue:
```diff
@@ -0,0 +1,86 @@
+}
+</script>
+
+<!-- eslint-disable max-len -->
```
Don't need this?

amanji replied:
No will remove

severinbeauvais replied:
Looks good. Resolving.

severinbeauvais commented on src/components/DigitalCredentials/dialogs/ConfirmCredentialsTermsofUseDialog.vue:
```diff
@@ -0,0 +1,105 @@
+          id="dialog-text"
+          class="warning-text"
+        >
+          I promise to be a good person and not use this credential for evil.
```
Seriously? lol

severinbeauvais replied:
There's more TOS coming here, right?

amanji replied:
Just placeholder text for now 😂

severinbeauvais replied:
OK 😆

amanji commented:
Still need to split the routes. Looking for the best way to do this

severinbeauvais commented on src/components/EntityInfo/EntityMenu.vue:
```diff
@@ -122,6 +98,28 @@
 
         <v-list dense>
           <v-list-item-group color="primary">
+            <!-- View/Add Digital Credentials -->
```
There's a height difference with this item.

![image](https://github.com/bcgov/business-filings-ui/assets/10002889/b9ec8589-0e6c-4a77-be2f-54cf0f76a6ba)

amanji replied:
Should be fixed now

severinbeauvais replied:
Yep. Resolving.

severinbeauvais commented on src/views/DigitalCredentials.vue:
```diff
@@ -1,213 +1,19 @@
+export default class DigitalCredentials extends Mixins(AllowableActionsMixin) {
+  mounted (): void {
+    if (!this.isAllowed(AllowableActions.DIGITAL_CREDENTIALS)) {
+      this.$router.push({ name: Routes.DASHBOARD })
```
I'm not sure if this is working -- I can't get to (or stay on) this page on the preview site.

https://business-filings-dev--pr-552-zpnetclq.web.app/BC0871407/

severinbeauvais replied:
Still not working using this URL:
https://business-filings-dev--pr-552-zpnetclq.web.app/FM1038931?accountid=2288

amanji replied:
I've updated the Mixin logic. It was matching on a route which will only work for the Entity Menu but not when routing to the Digital Credentials feature. This should now work.

severinbeauvais replied:
Yep, fixed now.

severinbeauvais commented on src/mixins/allowable-actions-mixin.ts:
```diff
@@ -92,10 +93,11 @@ export default class AllowableActionsMixin extends Vue {
       }
 
       case AllowableActions.DIGITAL_CREDENTIALS: {
+        // return true // *** FOR DEBUGGING ONLY
```
Remove if this is now working please.

severinbeauvais commented:
/gcbrun

bcregistry-sre commented:
Temporary Url for review: https://business-filings-dev--pr-552-zpnetclq.web.app

severinbeauvais commented:
/gcbrun

bcregistry-sre commented:
Temporary Url for review: https://business-filings-dev--pr-552-zpnetclq.web.app

severinbeauvais commented:
Oops, need to hide EntityMenu when user is in DC route.

![image](https://github.com/bcgov/business-filings-ui/assets/10002889/657f4d62-0772-4b65-a0d8-9ee344f6866f)

severinbeauvais commented:
@amanji Have you tested a bit on the preview site? Is it OK to merge or do you have more small fixes?

amanji commented:
Things look good. I have one more modal to go in

severinbeauvais commented:
/gcbrun

bcregistry-sre commented:
Temporary Url for review: https://business-filings-dev--pr-552-zpnetclq.web.app

severinbeauvais commented on src/components/DigitalCredentials/CredentialsStepper.vue:
```diff
@@ -121,7 +131,7 @@
           >
             <v-card-text
-              v-if="!issuedCredential"
+              v-if="!issuedCredential?.isIssued"
```
Caution: this will evaluate to True if issueCredential is falsy. Maybe compare with false or rework the statement?

PS It used to be that optional chaining didn't work in the template (since here it's JS code, not TS).

amanji replied:
Hmm, I think this is covered by the v-else above it.

severinbeauvais replied:
Maybe! So then you don't need the optional chaining here. And I hope nobody changes the logic in the template without catching this.

I also see the same issue on line 105.

severinbeauvais commented on src/components/DigitalCredentials/CredentialsStepper.vue:
```diff
@@ -252,7 +275,8 @@ export default class CredentialsStepper extends Vue {
   async issueCredential (credentialType: DigitalCredentialTypes): Promise<void> {
-    await LegalServices.sendCredentialOffer(this.getIdentifier, credentialType)
+    const { data: issuedCredential } = await LegalServices.sendCredentialOffer(this.getIdentifier, credentialType)
+    this.issuedCredential = issuedCredential
```
Is there any chance of `issuedCredential` being undefined (say, if sendCredentialOffer returns null)? That may affect the reactivity of this.issuedCredential.

amanji replied:
Added a default assignment to `null`

severinbeauvais replied:
Perfect, thanks

severinbeauvais commented on src/components/DigitalCredentials/dialogs/CredentialNotReceivedDialog.vue:
```diff
@@ -0,0 +1,131 @@
+              <p>Check your intenet connection</p>
+            </li>
+            <li>
+              <p>Try using your mobile data if available</p>
```
I've found that Vuetify attaches whitespace to paragraph elements, which I often override. If these list items look OK, just resolve this comment. Otherwise, I'd be tempted to remove the paragraph tags.

severinbeauvais replied:
Resolving this conversation. Instead, see other conversation below re: whitespace between text and list.

severinbeauvais commented on src/components/DigitalCredentials/dialogs/CredentialNotReceivedDialog.vue:
```diff
@@ -0,0 +1,131 @@
+          class="warning-text"
+        >
+          Credential offers can take up to a minute, depending on your internet connection. Try the following:
+          <ul>
```
Prob need some whitespace between the text and unordered list.

![image](https://github.com/bcgov/business-filings-ui/assets/10002889/aac0323e-a059-40bd-b83b-5d2782925cd6)

The spacing between the list items looks OK but I haven't seen your UI design.

amanji replied:
I think this is what's in the design. I confer with Kim/Hannah to see if we want to add some spacing

severinbeauvais replied:
Oops, I just noticed a typo above (intenet).

severinbeauvais replied:
Typos are fixed. Resolving and will leave it to Kim/Hannah to review.

severinbeauvais commented:
Akiff, I'm willing to merge this and deploy it to Dev, but is it working well enough to do that or should I wait?

amanji commented:
> Akiff, I'm willing to merge this and deploy it to Dev, but is it working well enough to do that or should I wait?

I think we can proceed with merging.

severinbeauvais reviewed (approved):
I approve to merge this PR, but please keep track of the remaining, unresolved conversations (for a future PR).
